File: app/src/main/java/uk/co/sundroid/activity/location/CoordsActivity.java
package uk.co.sundroid.activity.location;

import uk.co.sundroid.util.geo.Geocoder;
import uk.co.sundroid.R;
import uk.co.sundroid.domain.LocationDetails;
import uk.co.sundroid.util.location.LatitudeLongitude;
import uk.co.sundroid.util.SharedPrefsHelper;
import uk.co.sundroid.util.log.LogWrapper;
import android.app.Activity;
import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.TextView.OnEditorActionListener;
import android.widget.Toast;

public class CoordsActivity extends AbstractLocationActivity implements OnClickListener {
	
	private static final String TAG = CoordsActivity.class.getSimpleName();
	
	public static final int DIALOG_LOOKINGUP = 101;
	
    @Override
	protected int getLayout() {
		return R.layout.loc_coords;
	}

	@Override
	protected String getViewTitle() {
		return "Enter coordinates";
	}

    @Override
    public void onCreate(Bundle savedInstanceState) {
    	super.onCreate(savedInstanceState);
    	LogWrapper.d(TAG, "onCreate()");
		
    	View coordsSubmit = findViewById(R.id.coordsSubmit);
    	coordsSubmit.setOnClickListener(this);
    	
    	EditText coordsField = findViewById(R.id.coordsField);
    	coordsField.setOnEditorActionListener(new CoordsActionListener());
    	coordsField.addTextChangedListener(new TextWatcher() {
			public void afterTextChanged(Editable field) {
				try {
					parseLocation();
					show(R.id.coordsValid);
					hide(R.id.coordsInvalid);
				} catch (Exception e) {
					show(R.id.coordsInvalid);
					hide(R.id.coordsValid);
				}
			}
			public void beforeTextChanged(CharSequence string, int start, int count, int after) { }
			public void onTextChanged(CharSequence string, int start, int before, int count) { }
    	});
    }
    
	public void onClick(View button) {
		switch (button.getId()) {
			case R.id.coordsSubmit:
				startSubmit();
				return;
		}
        super.onClick(button);
	}
	
	private void startSubmit() {
		try {
			LatitudeLongitude location = parseLocation();
			startLookup(location);
		} catch (Exception e) {
			LogWrapper.e(TAG, "Parse error", e);
			Toast.makeText(getApplicationContext(), "Invalid coordinates. Please try again.", Toast.LENGTH_LONG).show();
		}
	}
	
	private LatitudeLongitude parseLocation() throws Exception {
		
		EditText coordsField = findViewById(R.id.coordsField);
		String coordsValue = coordsField.getText().toString().toUpperCase();
		
		LatitudeLongitude location;
		
		if (coordsValue.matches("[NS][0-9]+(\\.[0-9]+)? [WE][0-9]+(\\.[0-9]+)?")) {
			
			String lat = coordsValue.split(" ")[0];
			String lon = coordsValue.split(" ")[1];
			
			double latDbl = Double.parseDouble(lat.substring(1));
			double lonDbl = Double.parseDouble(lon.substring(1));
			if (lat.startsWith("S")) { latDbl = -latDbl; }
			if (lon.startsWith("W")) { lonDbl = -lonDbl; }
			
			location = new LatitudeLongitude(latDbl, lonDbl);
			
		} else if (coordsValue.matches("-?[0-9]+(\\.[0-9]+)? -?[0-9]+(\\.[0-9]+)?")) {
			
			String lat = coordsValue.split(" ")[0];
			String lon = coordsValue.split(" ")[1];
			double latDbl = Double.parseDouble(lat);
			double lonDbl = Double.parseDouble(lon);
			location = new LatitudeLongitude(latDbl, lonDbl);
			
		} else {
			
			location = new LatitudeLongitude(coordsValue);
			
		}
		
		return location;
		
	}

	private void startLookup(final LatitudeLongitude location) {
		LogWrapper.d(TAG, "startLookup()");
		
		showDialog(DIALOG_LOOKINGUP);
		final Activity activity = this;
		// FIXME AsyncTask
    	Thread thread = new Thread() {
    		@Override
    		public void run() {
    	    	final LocationDetails locationDetails = Geocoder.getLocationDetails(location, getApplicationContext());
    	    	dismissDialog(DIALOG_LOOKINGUP);
    			SharedPrefsHelper.saveSelectedLocation(activity, locationDetails);
    			if (locationDetails.getTimeZone() == null) {
    				Intent intent = new Intent(getApplicationContext(), TimeZonePickerActivity.class);
    				intent.putExtra(TimeZonePickerActivity.INTENT_MODE, TimeZonePickerActivity.MODE_SELECT);
    				startActivityForResult(intent, TimeZonePickerActivity.REQUEST_TIMEZONE);
    			} else {
    				setResult(LocationSelectActivity.RESULT_LOCATION_SELECTED);
    				finish();
    			}
    		}
    	};
    	thread.start();
    }
	
	@Override
	public Dialog onCreateDialog(int id) {
		if (id == DIALOG_LOOKINGUP) {
			ProgressDialog progressDialog = new ProgressDialog(this);
			progressDialog.setTitle("Looking up location...");
			progressDialog.setMessage("Looking up location, please wait. This can be disabled in settings.");
			progressDialog.setIndeterminate(true);
			progressDialog.setCancelable(false);
			return progressDialog;
		}
		return super.onCreateDialog(id);	
	}
	
	private class CoordsActionListener implements OnEditorActionListener {
		public boolean onEditorAction(TextView view, int id, KeyEvent arg2) {
			if (id == EditorInfo.IME_ACTION_GO) {
				startSubmit();
				return true;
			}
			return false;
		}
	}
    
}

File: app/src/main/kotlin/uk/co/sundroid/activity/location/CoordsActivity.kt
package uk.co.sundroid.activity.location

import uk.co.sundroid.util.location.Geocoder
import uk.co.sundroid.R
import uk.co.sundroid.util.location.LatitudeLongitude
import uk.co.sundroid.util.prefs.SharedPrefsHelper
import uk.co.sundroid.util.log.*
import android.app.Dialog
import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import android.widget.TextView
import android.widget.TextView.OnEditorActionListener
import android.widget.Toast

class CoordsActivity : AbstractLocationActivity() {

    override val layout: Int
        get() = R.layout.loc_coords

    override val viewTitle: String
        get() = "Enter coordinates"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val coordsSubmit = findViewById<View>(R.id.coordsSubmit)
        coordsSubmit.setOnClickListener({ _ -> startSubmit() })

        val coordsField = findViewById<EditText>(R.id.coordsField)
        coordsField.setOnEditorActionListener(CoordsActionListener())
        coordsField.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(field: Editable) {
                try {
                    parseLocation()
                    show(R.id.coordsValid)
                    hide(R.id.coordsInvalid)
                } catch (e: Exception) {
                    show(R.id.coordsInvalid)
                    hide(R.id.coordsValid)
                }
            }

            override fun beforeTextChanged(string: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(string: CharSequence, start: Int, before: Int, count: Int) {}
        })
    }

    private fun startSubmit() {
        try {
            val location = parseLocation()
            startLookup(location)
        } catch (e: Exception) {
            e(TAG, "Parse error", e)
            Toast.makeText(applicationContext, "Invalid coordinates. Please try again.", Toast.LENGTH_LONG).show()
        }

    }

    @Throws(Exception::class)
    private fun parseLocation(): LatitudeLongitude {

        val coordsField = findViewById<EditText>(R.id.coordsField)
        val coordsValue = coordsField.text.toString().toUpperCase()

        if (coordsValue.matches("[NS][0-9]+(\\.[0-9]+)? [WE][0-9]+(\\.[0-9]+)?".toRegex())) {

            val parts = coordsValue.split(" ")
            var latDbl = parts[0].substring(1).toDouble()
            var lonDbl = parts[1].substring(1).toDouble()
            if (parts[0].startsWith("S")) {
                latDbl = -latDbl
            }
            if (parts[1].startsWith("W")) {
                lonDbl = -lonDbl
            }
            return LatitudeLongitude(latDbl, lonDbl)

        } else if (coordsValue.matches("-?[0-9]+(\\.[0-9]+)? -?[0-9]+(\\.[0-9]+)?".toRegex())) {

            val parts = coordsValue.split(" ")
            return LatitudeLongitude(parts[0].toDouble(), parts[1].toDouble())

        } else {

            return LatitudeLongitude(coordsValue)

        }

    }

    private fun startLookup(location: LatitudeLongitude) {
        d(TAG, "startLookup()")

        showDialog(DIALOG_LOOKINGUP)
        val activity = this
        // FIXME AsyncTask
        val thread = object : Thread() {
            override fun run() {
                val locationDetails = Geocoder.getLocationDetails(location, applicationContext)
                dismissDialog(DIALOG_LOOKINGUP)
                SharedPrefsHelper.saveSelectedLocation(activity, locationDetails)
                if (locationDetails.timeZone == null) {
                    val intent = Intent(applicationContext, TimeZonePickerActivity::class.java)
                    intent.putExtra(TimeZonePickerActivity.INTENT_MODE, TimeZonePickerActivity.MODE_SELECT)
                    startActivityForResult(intent, TimeZonePickerActivity.REQUEST_TIMEZONE)
                } else {
                    setResult(LocationSelectActivity.RESULT_LOCATION_SELECTED)
                    finish()
                }
            }
        }
        thread.start()
    }

    public override fun onCreateDialog(id: Int): Dialog {
        if (id == DIALOG_LOOKINGUP) {
            val progressDialog = ProgressDialog(this)
            progressDialog.setTitle("Looking up location...")
            progressDialog.setMessage("Looking up location, please wait. This can be disabled in settings.")
            progressDialog.isIndeterminate = true
            progressDialog.setCancelable(false)
            return progressDialog
        }
        return super.onCreateDialog(id)
    }

    private inner class CoordsActionListener : OnEditorActionListener {
        override fun onEditorAction(view: TextView, id: Int, arg2: KeyEvent): Boolean {
            if (id == EditorInfo.IME_ACTION_GO) {
                startSubmit()
                return true
            }
            return false
        }
    }

    companion object {
        private val TAG = CoordsActivity::class.java.simpleName
        const val DIALOG_LOOKINGUP = 101
    }

}